File: iam-policy-lens/scripts/ts/analyzer.ts
/**
 * TypeScript Cloud Access Scanner
 * ===============================
 *
 * Problem Statement:
 * ------------------
 * This script performs static analysis on a TypeScript/Node.js codebase to identify Google Cloud, Vertex AI, GenAI, and Google API client library invocations.
 * The primary goal is to locate and extract these high-level method calls (e.g., `@google-cloud/storage.Bucket.file`), resolve their fully qualified names using the TypeScript TypeChecker, and determine credential provenance.
 *
 * How It Works:
 * -------------
 * 1. Project Loading: Automatically detects `tsconfig.json` or walks the target directory to load all TypeScript/JavaScript source files.
 * 2. AST & Type Checking: Uses `ts.createProgram` and `program.getTypeChecker()` to inspect CallExpressions and resolve symbols to fully qualified import paths.
 * 3. Credential Tracing: Traces client instantiation (e.g., `new Storage({ keyFilename: '...' })`) and variable declarations to classify credential provenance.
 * 4. JSON Output: Emits a structured JSON array conforming to `schema.json` to `stdout`, while logging progress to `stderr`.
 *
 * How to Run (from /Users/petrusca/Google/skills/iam-policy-lens):
 * ----------------------------------------------------------------
 * 1. Compile the TypeScript analyzer first (if not already built):
 *     npm --prefix scripts/ts run build
 *
 * 2. Run the analyzer passing the target project path as the first argument:
 *     node scripts/ts/dist/analyzer.js <path_to_project>
 *
 * 3. End-to-End Pipeline (Piping to Policy Generator):
 *     node scripts/ts/dist/analyzer.js /path/to/project | ./.venv/bin/python3 scripts/policy/policy.py
 * 
 * node scripts/ts/dist/analyzer.js ./../gcp_cost_optimizer_agent/ts
 */
import * as path from "path";
import { scanProject } from "./scanner.js";

function main() {
  const args = process.argv.slice(2);
  if (args.length < 1) {
    console.error("Usage:");
    console.error("  node scripts/ts/dist/analyzer.js <path_to_project>");
    process.exit(1);
  }

  const projectPath = path.resolve(args[0]);
  console.error(`Scanning: ${projectPath} for TypeScript/Node.js GAPIC calls`);

  const startTime = Date.now();
  const calls = scanProject(projectPath);
  const elapsed = (Date.now() - startTime) / 1000;

  if (calls.length > 0) {
    calls.sort((a, b) => {
      if (a.file_path === b.file_path) {
        return a.line - b.line;
      }
      return a.file_path.localeCompare(b.file_path);
    });

    console.log(JSON.stringify(calls, null, 2));
  } else {
    console.log("[]");
  }

  console.error(`\nScan completed in ${elapsed.toFixed(2)} seconds.`);
  console.error("====================================================");
}

main();

File: iam-policy-lens/scripts/go/analyzer.go
/*
Go Cloud Access Scanner
=======================

Problem Statement:
------------------
This script performs static analysis on a Go codebase to identify Google Cloud, Vertex AI, GenAI, and Google API client library invocations.
The primary goal is to locate and extract these high-level method calls (e.g., `cloud.google.com/go/storage.Client.Bucket`), resolve their fully qualified names using the Go TypeChecker, and determine credential provenance.

How It Works:
-------------
1. Project Loading: Uses `golang.org/x/tools/go/packages` to load all Go packages, syntax trees, and type information for the target directory.
2. AST & Type Checking: Inspects AST CallExpressions and resolves symbols to fully qualified import paths using `types.Info`.
3. Credential Tracing: Traces client instantiation (e.g., `storage.NewClient`) and variable declarations to classify credential provenance.
4. JSON Output: Emits a structured JSON array conforming to `schema.json` to `stdout`, while logging progress to `stderr`.

How to Run (from /Users/petrusca/Google/skills/iam-policy-lens):
----------------------------------------------------------------
1. Run the analyzer passing the target project path as the first argument:
    go run scripts/go/*.go <path_to_project>

2. End-to-End Pipeline (Piping to Policy Generator):
    go run scripts/go/*.go /path/to/project | ./.venv/bin/python3 scripts/policy/policy.py

go run scripts/go/*.go ./../gcp_cost_optimizer_agent/go
*/
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Usage:")
		fmt.Fprintln(os.Stderr, "  go run scripts/go/*.go <path_to_project>")
		os.Exit(1)
	}

	projectPath, err := filepath.Abs(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error resolving project path: %v\n", err)
		os.Exit(1)
	}

	fmt.Fprintf(os.Stderr, "Scanning: %s for Google Cloud Go API calls\n", projectPath)

	startTime := time.Now()

	callsChan, err := ScanProject(projectPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error scanning project: %v\n", err)
		os.Exit(1)
	}
	elapsed := time.Since(startTime)

	calls := collect(callsChan)
	if len(calls) > 0 {
		data, err := json.MarshalIndent(calls, "", "  ")
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error marshaling JSON: %v\n", err)
			os.Exit(1)
		}
		fmt.Println(string(data))
	} else {
		fmt.Println("[]")
	}

	fmt.Fprintf(os.Stderr, "\nScan completed in %.2f seconds.\n", elapsed.Seconds())
	fmt.Fprintln(os.Stderr, "====================================================")
}

func collect(ch <-chan GapicCall) []GapicCall {
	var calls []GapicCall
	for v := range ch {
		calls = append(calls, v)
	}
	sort.Slice(calls, func(i, j int) bool {
		if calls[i].FilePath == calls[j].FilePath {
			return calls[i].Line < calls[j].Line
		}
		return calls[i].FilePath < calls[j].FilePath
	})
	return calls
}
